File: game/projectilewars/scripts/vscripts/hero_selection.lua
require("typescript_lualib")
function ConfirmHeroPick(keys)
    if _G.options.AllowDuplicatePicksInRound and _G.pickedHeroes[keys.heroName] then
        if IsInToolsMode() then
            print("Duplicate hero picked in this round")
        end
        return false
    end
    if _G.options.AllowDuplicatePicksForPlayers and _G.pickedHeroes["player"..keys.playerID][keys.heroName] then
        if IsInToolsMode() then
            print("Duplicate hero picked for player")
        end
        return false
    end
    StoreHeroPick(keys.heroName,keys.PlayerID)
    return true
end
function StoreHeroPick(heroname,pid)
    _G.pickedHeroes[heroname]=true
    _G.pickedHeroes[pid][heroname]=true
    local hero = CreateHeroForPlayer(heroname,CDOTA_PlayerResource.GetPlayer(PlayerResource,pid))

    UTIL_Remove(hero)
end
function GetRandomHero(pid)
    local heroes = LoadKeyValues("scripts/npc/npc_heroes.txt")

    local rand = RandomInt(0,#heroes-1)

    local heroname = heroes[rand+1]

    local t = {heroName=heroname,PlayerID=pid}

    local b = ConfirmHeroPick(t)

    while not b do
        rand=RandomInt(0,#heroes-1)
        heroname=heroes[rand+1]
        t={heroName=heroname,PlayerID=pid}
    end
end
function HeroPickPhaseStarted()
    _G.pickedHeroes={}
    for i=-1,23,1 do
        _G.pickedHeroes["player"..i]=(_G.pickedHeroes["player"..i] or {})
    end
end
function HeroPickPhaseEnded()
    for i=1,DOTA_MAX_TEAM,1 do
        if CDOTA_PlayerResource.IsValidTeamPlayerID(PlayerResource,i) and not CDOTA_PlayerResource.GetSelectedHeroEntity(PlayerResource,i) then
        end
    end
end

File: game/projectilewars/scripts/vscripts/hero_selection.ts

function ConfirmHeroPick(keys):boolean {
  if (_G.options.AllowDuplicatePicksInRound && _G.pickedHeroes[keys.heroName] ) {
    if (IsInToolsMode()) {print("Duplicate hero picked in this round")}
    return false
  }
  if (_G.options.AllowDuplicatePicksForPlayers && _G.pickedHeroes["player"+keys.playerID][keys.heroName] ) {
    if (IsInToolsMode()) {print("Duplicate hero picked for player")}
    return false
  }
  StoreHeroPick(keys.heroName,keys.PlayerID)
  return true
}


function StoreHeroPick(heroname,pid:PlayerID):void {
  _G.pickedHeroes[heroname] = true
  _G.pickedHeroes[pid][heroname] = true
  let hero = CreateHeroForPlayer(heroname,PlayerResource.GetPlayer(pid))
  UTIL_Remove(hero)
}

function GetRandomHero(pid:PlayerID) {
  let heroes = LoadKeyValues("scripts/npc/npc_heroes.txt") as string[]
  let rand = RandomInt(0,heroes.length-1)
  let heroname = heroes[rand]
  let t = {heroName:heroname,PlayerID:pid}
  let b = ConfirmHeroPick(t)
  while(!b) {
    rand = RandomInt(0,heroes.length-1)
    heroname = heroes[rand]
    t = {heroName:heroname,PlayerID:pid}
  }
}

function HeroPickPhaseStarted() {
  _G.pickedHeroes = {}

  for (let i=-1;i<=23;i++) {
    _G.pickedHeroes["player"+i] = _G.pickedHeroes["player"+i] || {}
  }
}

function HeroPickPhaseEnded() {
  // Random for all players
  for (let i =1 as PlayerID;i<=DOTALimits_t.DOTA_MAX_TEAM;i++) {
    if (PlayerResource.IsValidTeamPlayerID(i) && !PlayerResource.GetSelectedHeroEntity(i)) {

    }
  }
}